// @skbkontur/react-ui/components/DatePickerOld/DateInput.js
"use strict";exports.__esModule = true;exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));

var _Input = _interopRequireDefault(require("../Input"));

var _filterProps = _interopRequireDefault(require("../filterProps"));
var _ensureOldIEClassName = require("../ensureOldIEClassName");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { "default": obj };}function _getRequireWildcardCache() {if (typeof WeakMap !== "function") return null;var cache = new WeakMap();_getRequireWildcardCache = function _getRequireWildcardCache() {return cache;};return cache;}function _interopRequireWildcard(obj) {if (obj && obj.__esModule) {return obj;}if (obj === null || typeof obj !== "object" && typeof obj !== "function") {return { "default": obj };}var cache = _getRequireWildcardCache();if (cache && cache.has(obj)) {return cache.get(obj);}var newObj = {};var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var key in obj) {if (Object.prototype.hasOwnProperty.call(obj, key)) {var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;if (desc && (desc.get || desc.set)) {Object.defineProperty(newObj, key, desc);} else {newObj[key] = obj[key];}}}newObj["default"] = obj;if (cache) {cache.set(obj, newObj);}return newObj;}function _extends() {_extends = Object.assign || function (target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i];for (var key in source) {if (Object.prototype.hasOwnProperty.call(source, key)) {target[key] = source[key];}}}return target;};return _extends.apply(this, arguments);}function _assertThisInitialized(self) {if (self === void 0) {throw new ReferenceError("this hasn't been initialised - super() hasn't been called");}return self;}function _inheritsLoose(subClass, superClass) {subClass.prototype = Object.create(superClass.prototype);subClass.prototype.constructor = subClass;subClass.__proto__ = superClass;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}

var DateBlocks = [{ index: 0, start: 0, end: 2 }, { index: 1, start: 3, end: 5 }, { index: 2, start: 6, end: 10 }];

var INPUT_PASS_PROPS = {
  autoFocus: true,
  disabled: true,
  warning: true,
  error: true,
  size: true,
  placeholder: true,

  onInput: true,
  onKeyDown: true,
  onKeyPress: true,
  onKeyUp: true,
  onMouseEnter: true,
  onMouseLeave: true,
  onMouseOver: true };

























var isIE8 = _ensureOldIEClassName.ieVerison === 8;var

DateInput = /*#__PURE__*/function (_Component) {_inheritsLoose(DateInput, _Component);function DateInput() {var _this;for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {args[_key] = arguments[_key];}_this = _Component.call.apply(_Component, [this].concat(args)) || this;_defineProperty(_assertThisInitialized(_this), "_input", void 0);_defineProperty(_assertThisInitialized(_this), "_focused",














    false);_defineProperty(_assertThisInitialized(_this), "_icon", void 0);_defineProperty(_assertThisInitialized(_this), "_cursorPosition",

    0);_defineProperty(_assertThisInitialized(_this), "_setCursorPosition",




























    function (cursorPosition) {
      _this._input && _this._input.setSelectionRange(cursorPosition, cursorPosition);
    });_defineProperty(_assertThisInitialized(_this), "_selectCurrentBlock",

    function (input) {
      if (trim(_this.props.value) === '') {
        _this._selectBlock(_this._getSelectedBlock(0));
        return;
      }var _getInputSelection =
      getInputSelection(input),start = _getInputSelection.start,end = _getInputSelection.end;
      if (start !== end) {
        return;
      }
      _this._selectBlock(_this._getSelectedBlock(start));
    });_defineProperty(_assertThisInitialized(_this), "_handleClick",

    function (event) {
      if (_this._focused) {
        // $FlowIgnore
        _this._selectCurrentBlock(event.target);
      }
    });_defineProperty(_assertThisInitialized(_this), "_handleKeyDown",

    function (event) {
      if (_this._checkIfBadKeyDownEvent(event)) {
        return;
      }
      if (_this._isVerticalArrows(event)) {
        event.preventDefault();
        _this._handleVerticalKey(event);
      }
      if (!isIE8 && _this._isHorizontalArrows(event)) {
        event.preventDefault();
        _this._moveSelectionBlock(event);
      }
      if (_this._isSeparatorKey(event)) {
        event.preventDefault();
        _this._handleSeparatorKey(event);
      }
      if (event.key === 'Enter') {
        event.preventDefault();
        _this.props.onSubmit && _this.props.onSubmit();
      }
    });_defineProperty(_assertThisInitialized(_this), "_handleInputPaste",

    function (event) {
      var text = event.clipboardData.getData('text').trim();
      var re = /([\d\_]{1,2}).([\d\_]{1,2}).([\d\_]{4})/;
      var execData = re.exec(text);
      if (!execData) {
        return;
      }var
      day = execData[1],month = execData[2],year = execData[3];
      var nextValue = [day, month, year].
      filter(Boolean).
      map(pad2).
      join('.');

      if (_this.props.onChange) {
        _this.props.onChange(nextValue);
      }
      setTimeout(function () {
        _this._input && _this._input.setSelectionRange(0, 12);
      }, 100);
    });_defineProperty(_assertThisInitialized(_this), "_checkIfBadKeyDownEvent",

    function (event) {
      var AllowedKeys = ['Enter', 'ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight', '.', ',', ' '];
      return !AllowedKeys.includes(event.key);
    });_defineProperty(_assertThisInitialized(_this), "_isVerticalArrows",

    function (event) {
      return event.key === 'ArrowUp' || event.key === 'ArrowDown';
    });_defineProperty(_assertThisInitialized(_this), "_isHorizontalArrows",

    function (event) {
      return event.key === 'ArrowLeft' || event.key === 'ArrowRight';
    });_defineProperty(_assertThisInitialized(_this), "_isSeparatorKey",

    function (event) {
      return [' ', ',', '.'].includes(event.key);
    });_defineProperty(_assertThisInitialized(_this), "_moveSelectionBlock",

    function (event) {var _getInputSelection2 =
      getInputSelection(event.currentTarget),start = _getInputSelection2.start;
      var currentSelectedBlock = _this._getSelectedBlock(start);
      var step = event.key === 'ArrowLeft' ? -1 : 1;
      var nextSelectedBlockIndex = Math.max(Math.min(currentSelectedBlock.index + step, 2), 0);
      _this._selectBlock(DateBlocks[nextSelectedBlockIndex]);
    });_defineProperty(_assertThisInitialized(_this), "_selectBlock",

    function (_ref) {var index = _ref.index,start = _ref.start,end = _ref.end;
      setTimeout(function () {
        _this._input && _this._input.setSelectionRange(start, end);
      }, 10);
    });_defineProperty(_assertThisInitialized(_this), "_handleSeparatorKey",

    function (event) {
      var re = /([\d\_]{0,2}).?([\d\_]{0,2}).?([\d\_]{0,4})/;var _getInputSelection3 =
      getInputSelection(event.currentTarget),start = _getInputSelection3.start;
      if (start !== 1 && start !== 4) {
        return;
      }var _re$exec =
      re.exec(_this.props.value.replace('_', '')),_re$exec$ = _re$exec[1],day = _re$exec$ === void 0 ? '' : _re$exec$,_re$exec$2 = _re$exec[2],month = _re$exec$2 === void 0 ? '' : _re$exec$2,_re$exec$3 = _re$exec[3],year = _re$exec$3 === void 0 ? '' : _re$exec$3;

      var nextValue = [day, month, year].
      filter(Boolean).
      map(pad2).
      join('.');

      if (nextValue.length < 7) {
        nextValue += '.';
      }

      if (_this.props.onChange) {
        _this.props.onChange(nextValue);
      }
      _this._moveSelectionBlock(event);
    });_defineProperty(_assertThisInitialized(_this), "_getSelectedBlock",

    function (cursorPosition) {
      return cursorPosition < 3 ? DateBlocks[0] : cursorPosition < 6 ? DateBlocks[1] : DateBlocks[2];
    });_defineProperty(_assertThisInitialized(_this), "_handleChange",

    function (event, value) {
      var isBackspace = trim(_this.props.value).length >= trim(value).length;

      _this.props.onChange(value);var

      currentTarget = event.currentTarget;
      if (!isBackspace && !isIE8) {
        setTimeout(function () {
          var start = getInputSelection(currentTarget).start;
          if (start === 3 || start === 6) {
            _this._selectBlock(_this._getSelectedBlock(start));
          }
        }, 100);
      }
    });_defineProperty(_assertThisInitialized(_this), "_handleVerticalKey",

    function (event) {
      event.preventDefault();

      var dateValue = event.currentTarget.value;

      var isValid = /\d{2}.\d{2}.\d{4}/.test(dateValue);
      if (!isValid) {
        return;
      }var _getInputSelection4 =

      getInputSelection(event.currentTarget),start = _getInputSelection4.start;
      var selectedBlock = _this._getSelectedBlock(start);

      var step = 0;
      if (event.key === 'ArrowUp') {
        step = 1;
      }
      if (event.key === 'ArrowDown') {
        step = -1;
      }var _dateValue$split =

      dateValue.split('.'),day = _dateValue$split[0],month = _dateValue$split[1],year = _dateValue$split[2];
      day = Number(day);
      month = Number(month) - 1;
      year = Number(year);

      var date;
      switch (selectedBlock.index) {
        case 0:
          date = new Date(Date.UTC(year, month, day + step));
          break;
        case 1:
          date = new Date(Date.UTC(year, month + step, day));
          break;
        case 2:
          date = new Date(Date.UTC(year + step, month, day));
          break;
        default:
          throw new TypeError('Unexpected block index ' + selectedBlock.index);}


      var newDay = date.getUTCDate();
      var newMonth = date.getUTCMonth() + 1;
      var newYear = date.getUTCFullYear();
      var newDate = [newDay, newMonth, newYear].
      filter(Boolean).
      map(pad2).
      join('.');

      _this.props.onChange(newDate);
      _this._selectBlock(selectedBlock);
    });_defineProperty(_assertThisInitialized(_this), "_handleFocus",

    function (event) {
      _this._focused = true;
      if (_this.props.onFocus) {
        _this.props.onFocus();
      }
    });_defineProperty(_assertThisInitialized(_this), "_handleBlur",

    function (e) {
      _this._focused = false;
      if (_this.props.onBlur) {
        _this.props.onBlur(e);
      }
    });_defineProperty(_assertThisInitialized(_this), "getInputRef",

    function (ref) {
      _this._input = ref;
    });_defineProperty(_assertThisInitialized(_this), "getIconRef",

    function (ref) {
      _this._icon = ref;
    });return _this;}var _proto = DateInput.prototype;_proto.render = function render() {var maskChar = this.props.withMask ? '_' : null;return _react["default"].createElement(_Input["default"], _extends({}, (0, _filterProps["default"])(this.props, INPUT_PASS_PROPS), { mask: "99.99.9999", maskChar: maskChar, maxLength: 12, value: this.props.value, width: "100%", onBlur: this._handleBlur, onFocus: this._handleFocus, onChange: this._handleChange, onMouseUp: this._handleClick, onKeyDown: this._handleKeyDown, onPaste: this._handleInputPaste, ref: this.getInputRef }));};_proto.componentDidMount = function componentDidMount() {if (this.props.getInputRef && this._input) {this.props.getInputRef(this._input);}};return DateInput;}(_react.Component);exports["default"] = DateInput;_defineProperty(DateInput, "__KONTUR_REACT_UI__", 'DateInputOld');_defineProperty(DateInput, "propTypes", { getInputRef: _propTypes["default"].func, opened: _propTypes["default"].bool.isRequired, placeholder: _propTypes["default"].string, size: _propTypes["default"].oneOf(['small', 'medium', 'large']), value: _propTypes["default"].string.isRequired, onBlur: _propTypes["default"].func, onChange: _propTypes["default"].func.isRequired, onFocus: _propTypes["default"].func });


var pad2 = function pad2(v) {return v.toString().padStart(2, '0');};

function trim(str) {
  return str.replace(/[\_\.]/g, '');
}

/**
   * imported from http://stackoverflow.com/questions/4928586/ddg#4931963
   */
function getInputSelection(el) {
  var start = 0;
  var end = 0;

  if (typeof el.selectionStart === 'number' && typeof el.selectionEnd === 'number') {
    start = el.selectionStart;
    end = el.selectionEnd;
  } else if (document.selection) {
    // eslint-disable-next-line flowtype/no-weak-types
    var range = document.selection.createRange();

    if (range && range.parentElement() === el) {
      var len = el.value.length;
      var normalizedValue = el.value.replace(/\r\n/g, '\n');

      // Create a working TextRange that lives only in the input
      var textInputRange = el.createTextRange();
      textInputRange.moveToBookmark(range.getBookmark());

      // Check if the start and end of the selection are at the very end
      // of the input, since moveStart/moveEnd doesn't return what we want
      // in those cases
      var endRange = el.createTextRange();
      endRange.collapse(false);

      if (textInputRange.compareEndPoints('StartToEnd', endRange) > -1) {
        start = end = len;
      } else {
        start = -textInputRange.moveStart('character', -len);
        start += normalizedValue.slice(0, start).split('\n').length - 1;

        if (textInputRange.compareEndPoints('EndToEnd', endRange) > -1) {
          end = len;
        } else {
          end = -textInputRange.moveEnd('character', -len);
          end += normalizedValue.slice(0, end).split('\n').length - 1;
        }
      }
    }
  }

  return {
    start: start,
    end: end };

}
