// @skbkontur/react-ui/components/MenuItem/MenuItem.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Emotion_1 = require("../../lib/theming/Emotion");
var Mixins_1 = require("../../lib/styles/Mixins");
var jsStyles = {
  root: function root(t) {
    return Emotion_1.css(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      ", ";\n\n      line-height: 18px;\n      padding: 6px 18px 7px 8px;\n    "], ["\n      ", ";\n\n      line-height: 18px;\n      padding: 6px 18px 7px 8px;\n    "])), Mixins_1.resetButton());
  },
  hover: function hover(t) {
    // Color with !important in purpose to override `a:hover`
    return Emotion_1.css(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n      background: ", ";\n      color: ", " !important;\n    "], ["\n      background: ", ";\n      color: ", " !important;\n    "])), t.dropdownMenuHoverBg, t.textColorInvert);
  },
  selected: function selected(t) {
    return Emotion_1.css(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n      background: ", ";\n    "], ["\n      background: ", ";\n    "])), t.dropdownMenuSelectedBg);
  },
  disabled: function disabled(t) {
    return Emotion_1.css(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n      color: ", ";\n    "], ["\n      color: ", ";\n    "])), t.textColorDisabled);
  },
  link: function link(t) {
    return Emotion_1.css(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n      color: ", ";\n    "], ["\n      color: ", ";\n    "])), t.linkColor);
  },
  withIcon: function withIcon(t) {
    return Emotion_1.css(templateObject_6 || (templateObject_6 = tslib_1.__makeTemplateObject(["\n      padding-left: ", ";\n    "], ["\n      padding-left: ", ";\n    "])), t.menuItemPaddingForIcon);
  } };

exports["default"] = jsStyles;
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;
