// @skbkontur/react-ui/components/Fias/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EstateStatuses;
(function (EstateStatuses) {
  EstateStatuses["Hold"] = "Hold";
  EstateStatuses["House"] = "House";
  EstateStatuses["HouseHold"] = "HouseHold";
  EstateStatuses["None"] = "None";
})(EstateStatuses = exports.EstateStatuses || (exports.EstateStatuses = {}));
var StructureStatuses;
(function (StructureStatuses) {
  StructureStatuses["Structure"] = "Structure";
  StructureStatuses["Construction"] = "Construction";
  StructureStatuses["Liter"] = "Liter";
  StructureStatuses["None"] = "None";
})(StructureStatuses = exports.StructureStatuses || (exports.StructureStatuses = {}));
var LiveStatuses;
(function (LiveStatuses) {
  LiveStatuses["active"] = "active";
  LiveStatuses["inactive"] = "inactive";
})(LiveStatuses = exports.LiveStatuses || (exports.LiveStatuses = {}));
var Fields;
(function (Fields) {
  Fields["region"] = "region";
  Fields["district"] = "district";
  Fields["city"] = "city";
  Fields["intracityarea"] = "intracityarea";
  Fields["settlement"] = "settlement";
  Fields["planningstructure"] = "planningstructure";
  Fields["street"] = "street";
  Fields["stead"] = "stead";
  Fields["house"] = "house";
  Fields["room"] = "room";
})(Fields = exports.Fields || (exports.Fields = {}));
var ExtraFields;
(function (ExtraFields) {
  ExtraFields["postalcode"] = "postalcode";
})(ExtraFields = exports.ExtraFields || (exports.ExtraFields = {}));
var FormValidation;
(function (FormValidation) {
  FormValidation["Error"] = "Error";
  FormValidation["Warning"] = "Warning";
  FormValidation["None"] = "None";
})(FormValidation = exports.FormValidation || (exports.FormValidation = {}));
