// @skbkontur/react-ui/lib/LayoutEvents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fbemitter_1 = require("fbemitter");
var react_dom_1 = require("react-dom");
var emitterCache;
function getEmitter() {
  if (!emitterCache) {
    emitterCache = new fbemitter_1.EventEmitter();
  }
  return emitterCache;
}
function listenBrowserEvents() {
  window.addEventListener('scroll', emit);
  window.addEventListener('resize', emit);
}
function unlistenBrowserEvents() {
  window.removeEventListener('scroll', emit);
  window.removeEventListener('resize', emit);
}
function addListener(callback) {
  var emitter = getEmitter();
  if (emitter.listeners('layout').length === 0) {
    listenBrowserEvents();
  }
  var token = emitter.addListener('layout', callback);
  return {
    remove: function remove() {
      token.remove();
      if (emitter.listeners('layout').length === 0) {
        unlistenBrowserEvents();
      }
    } };

}
function emit() {
  react_dom_1.unstable_batchedUpdates(function () {
    getEmitter().emit('layout');
  });
}
exports["default"] = {
  addListener: addListener,
  emit: emit };
