// @skbkontur/react-ui/components/Kladr/AddressModal.js
"use strict";exports.__esModule = true;exports["default"] = void 0;

var React = _interopRequireWildcard(require("react"));

var _Button = _interopRequireDefault(require("../Button"));
var _ComboBoxOld = _interopRequireDefault(require("../ComboBoxOld"));
var _Gapped = _interopRequireDefault(require("../Gapped"));
var _Input = _interopRequireDefault(require("../Input"));
var _Modal = _interopRequireDefault(require("../Modal"));

var _KladrAPI = require("./KladrAPI");

var util = _interopRequireWildcard(require("./util"));var _AddressModalModule = _interopRequireDefault(require("./AddressModal.module.css"));function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { "default": obj };}function _getRequireWildcardCache() {if (typeof WeakMap !== "function") return null;var cache = new WeakMap();_getRequireWildcardCache = function _getRequireWildcardCache() {return cache;};return cache;}function _interopRequireWildcard(obj) {if (obj && obj.__esModule) {return obj;}if (obj === null || typeof obj !== "object" && typeof obj !== "function") {return { "default": obj };}var cache = _getRequireWildcardCache();if (cache && cache.has(obj)) {return cache.get(obj);}var newObj = {};var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var key in obj) {if (Object.prototype.hasOwnProperty.call(obj, key)) {var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;if (desc && (desc.get || desc.set)) {Object.defineProperty(newObj, key, desc);} else {newObj[key] = obj[key];}}}newObj["default"] = obj;if (cache) {cache.set(obj, newObj);}return newObj;}function _extends() {_extends = Object.assign || function (target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i];for (var key in source) {if (Object.prototype.hasOwnProperty.call(source, key)) {target[key] = source[key];}}}return target;};return _extends.apply(this, arguments);}function ownKeys(object, enumerableOnly) {var keys = Object.keys(object);if (Object.getOwnPropertySymbols) {var symbols = Object.getOwnPropertySymbols(object);if (enumerableOnly) symbols = symbols.filter(function (sym) {return Object.getOwnPropertyDescriptor(object, sym).enumerable;});keys.push.apply(keys, symbols);}return keys;}function _objectSpread(target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i] != null ? arguments[i] : {};if (i % 2) {ownKeys(Object(source), true).forEach(function (key) {_defineProperty(target, key, source[key]);});} else if (Object.getOwnPropertyDescriptors) {Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));} else {ownKeys(Object(source)).forEach(function (key) {Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));});}}return target;}function _assertThisInitialized(self) {if (self === void 0) {throw new ReferenceError("this hasn't been initialised - super() hasn't been called");}return self;}function _inheritsLoose(subClass, superClass) {subClass.prototype = Object.create(superClass.prototype);subClass.prototype.constructor = subClass;subClass.__proto__ = superClass;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}





































var PLACES = {
  '0': 'index',
  '1': 'region',
  '2': 'district',
  '4': 'city',
  '8': 'settlement',
  '16': 'street' };var


AddressModal = /*#__PURE__*/function (_React$Component) {_inheritsLoose(AddressModal, _React$Component);











  function AddressModal(props) {var _this;
    _this = _React$Component.call(this, props) || this;_defineProperty(_assertThisInitialized(_this), "_regionProps", void 0);_defineProperty(_assertThisInitialized(_this), "_districtProps", void 0);_defineProperty(_assertThisInitialized(_this), "_cityProps", void 0);_defineProperty(_assertThisInitialized(_this), "_settlementProps", void 0);_defineProperty(_assertThisInitialized(_this), "_streetProps", void 0);_defineProperty(_assertThisInitialized(_this), "_houseProps", void 0);_defineProperty(_assertThisInitialized(_this), "_buildingProps", void 0);_defineProperty(_assertThisInitialized(_this), "_roomProps", void 0);_defineProperty(_assertThisInitialized(_this), "_verifyPromise", void 0);_defineProperty(_assertThisInitialized(_this), "_handleSave",





























































































































































































































































    function () {
      _this.props.onChange({ address: _this.state.address });
      _this.props.onClose();
    });_this.state = { address: props.address || {}, invalidField: null };_this._regionProps = _this.createFieldProps('region', [], 'Region');_this._districtProps = _this.createFieldProps('district', ['region'], 'District');_this._cityProps = _this.createFieldProps('city', ['region', 'district'], 'City');_this._settlementProps = _this.createFieldProps('settlement', ['region', 'district', 'city'], 'Settlement');_this._streetProps = _this.createFieldProps('street', ['region', 'district', 'city', 'settlement'], 'Street');_this._houseProps = _this.createSimpleFieldProps('house');_this._buildingProps = _this.createSimpleFieldProps('building');_this._roomProps = _this.createSimpleFieldProps('room');return _this;}var _proto = AddressModal.prototype;_proto.createFieldProps = function createFieldProps(field, parents, level) {return { source: this.createSource(field, parents, level), onChange: this.createHandler(field), renderItem: this._renderItem.bind(this, field, parents), renderValue: renderValue.bind(null, field), valueToString: renderValue.bind(null, field), recover: recover };};_proto.createSimpleFieldProps = function createSimpleFieldProps(field) {var _this2 = this;return { onChange: function onChange(event, value) {var _objectSpread2;var address = _objectSpread({}, _this2.state.address, (_objectSpread2 = {}, _objectSpread2[field] = value, _objectSpread2));_this2.setState({ address: address });} };};_proto.createHandler = function createHandler(place) {var _this3 = this;return function (e, value, info) {var address = _objectSpread({}, _this3.state.address);if (info) {for (var item in PLACES) {var itemPlace = PLACES[item];if (itemPlace !== place && info.address[itemPlace]) {address[itemPlace] = info.address[itemPlace];} else if (itemPlace === place) {address[place] = value;break;}}} else if (value.name) {address[place] = value;} else {address[place] = null;}_this3.setState({ address: address });_this3.check(address);};};_proto.check = function check(address) {var _this4 = this;var promise = (0, _KladrAPI.verify)(address);this._verifyPromise = promise;promise.then(function (result) {if (promise !== _this4._verifyPromise) {return;}_this4._verifyPromise = null;var invalidField = null;if (!result.isKladrAddress) {invalidField = PLACES[result.invalidItem];}_this4.setState({ address: _objectSpread({}, _this4.state.address, {}, result.address), invalidField: invalidField });});};_proto.createSource = function createSource(field, parents, level) {var _this5 = this;return function (searchText) {var parentCode = null;for (var _iterator = parents, _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : _iterator[Symbol.iterator]();;) {var _ref;if (_isArray) {if (_i >= _iterator.length) break;_ref = _iterator[_i++];} else {_i = _iterator.next();if (_i.done) break;_ref = _i.value;}var parentName = _ref;var parent = _this5.state.address[parentName];if (parent && typeof parent === 'object' && parent.code) {parentCode = parent.code;}}return (0, _KladrAPI.search)(searchText, "[" + level + "]", parentCode).then(function (values) {return { values: values.map(function (address) {return address[field];}), infos: values.map(function (address) {return { searchText: searchText, address: address };}) };});};};_proto.setStateAddress = function setStateAddress(key, value) {var _objectSpread3;this.setState({ address: _objectSpread({}, this.state.address, (_objectSpread3 = {}, _objectSpread3[key] = value, _objectSpread3)) });};_proto.render = function render() {return React.createElement(_Modal["default"], { width: 520, onClose: this.props.onClose }, React.createElement(_Modal["default"].Header, null, this.props.title), React.createElement(_Modal["default"].Body, null, this._renderForm()), React.createElement(_Modal["default"].Footer, { panel: true }, React.createElement(_Gapped["default"], null, React.createElement(_Button["default"], { size: "medium", use: "primary", onClick: this._handleSave }, "\u0421\u043E\u0445\u0440\u0430\u043D\u0438\u0442\u044C"), React.createElement(_Button["default"], { size: "medium", onClick: this.props.onClose }, "\u041E\u0442\u043C\u0435\u043D\u0430"))));};_proto._renderForm = function _renderForm() {var _this6 = this;var invalidField = this.state.invalidField;return React.createElement(_Gapped["default"], { vertical: true }, React.createElement("div", { className: _AddressModalModule["default"].row }, React.createElement("div", { className: _AddressModalModule["default"].label }, "\u0418\u043D\u0434\u0435\u043A\u0441"), React.createElement("div", { className: _AddressModalModule["default"].field }, React.createElement(_Input["default"], { value: this.state.address.index, width: "100%", onChange: function onChange(e) {_this6.setStateAddress('index', e.target.value);} }))), React.createElement("div", { className: _AddressModalModule["default"].row }, React.createElement("div", { className: _AddressModalModule["default"].label }, "\u0420\u0435\u0433\u0438\u043E\u043D"), React.createElement("div", { className: _AddressModalModule["default"].field }, React.createElement(_ComboBoxOld["default"], _extends({ value: this.state.address.region, error: invalidField === 'region', placeholder: "\u041C\u043E\u0436\u043D\u043E \u0432\u0432\u043E\u0434\u0438\u0442\u044C \u043A\u043E\u0434 \u0438\u043B\u0438 \u043D\u0430\u0437\u0432\u0430\u043D\u0438\u0435", width: "100%" }, this._regionProps)))), React.createElement("div", { className: _AddressModalModule["default"].row }, React.createElement("div", { className: _AddressModalModule["default"].label }, "\u0420\u0430\u0439\u043E\u043D"), React.createElement("div", { className: _AddressModalModule["default"].field }, React.createElement(_ComboBoxOld["default"], _extends({ value: this.state.address.district, error: invalidField === 'district', width: "100%" }, this._districtProps)))), React.createElement("div", { className: _AddressModalModule["default"].row }, React.createElement("div", { className: _AddressModalModule["default"].label }, "\u0413\u043E\u0440\u043E\u0434"), React.createElement("div", { className: _AddressModalModule["default"].field }, React.createElement(_ComboBoxOld["default"], _extends({ value: this.state.address.city, error: invalidField === 'city', width: "100%" }, this._cityProps)))), React.createElement("div", { className: _AddressModalModule["default"].row }, React.createElement("div", { className: _AddressModalModule["default"].label }, "\u041D\u0430\u0441\u0435\u043B\u0435\u043D\u043D\u044B\u0439 \u043F\u0443\u043D\u043A\u0442"), React.createElement("div", { className: _AddressModalModule["default"].field }, React.createElement(_ComboBoxOld["default"], _extends({ value: this.state.address.settlement, error: invalidField === 'settlement', width: "100%" }, this._settlementProps)))), React.createElement("div", { className: _AddressModalModule["default"].row }, React.createElement("div", { className: _AddressModalModule["default"].label }, "\u0423\u043B\u0438\u0446\u0430"), React.createElement("div", { className: _AddressModalModule["default"].field }, React.createElement(_ComboBoxOld["default"], _extends({ value: this.state.address.street, error: invalidField === 'street', width: "100%" }, this._streetProps)))), React.createElement("div", { className: _AddressModalModule["default"].row }, React.createElement("div", { className: _AddressModalModule["default"].label }, "\u0414\u043E\u043C"), React.createElement("div", { className: _AddressModalModule["default"].field }, React.createElement(_Input["default"], _extends({ value: this.state.address.house, width: 100 }, this._houseProps)))), React.createElement("div", { className: _AddressModalModule["default"].row }, React.createElement("div", { className: _AddressModalModule["default"].label }, "\u041A\u043E\u0440\u043F\u0443\u0441"), React.createElement("div", { className: _AddressModalModule["default"].field }, React.createElement(_Input["default"], _extends({ value: this.state.address.building || '', width: 100 }, this._buildingProps)))), React.createElement("div", { className: _AddressModalModule["default"].row }, React.createElement("div", { className: _AddressModalModule["default"].label }, "\u041A\u0432\u0430\u0440\u0442\u0438\u0440\u0430 / \u043E\u0444\u0438\u0441"), React.createElement("div", { className: _AddressModalModule["default"].field }, React.createElement(_Input["default"], _extends({ value: this.state.address.room, width: "100px" }, this._roomProps)))));};_proto._renderItem = function _renderItem(field, parents, place, info) {var parentNames = [];for (var i = parents.length - 1; i >= 0; --i) {var parentField = parents[i];if (this.state.address[parentField]) {break;}var parent = info.address[parentField];if (parent && typeof parent === 'object') {var parentName = parent.name;if (parentField === 'region') {parentName = parent.code.substr(0, 2) + ' ' + parentName;}parentNames.unshift(parentName);}}return React.createElement("div", null, util.placeName(place, field), parentNames.length > 0 && React.createElement("div", { className: _AddressModalModule["default"].menuItemParents }, parentNames.join(', ')));};return AddressModal;}(React.Component);exports["default"] = AddressModal;


function renderValue(type, place) {
  return place ? util.placeName(place, type) : 'null';
}

function recover(searchText) {
  return {
    value: { name: searchText } };

}
