// @skbkontur/react-ui/components/ZIndex/ZIndexStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var componentPriorities = {
  Toast: 10,
  Sidepage: 9,
  Modal: 9,
  Sticky: 7,
  DropdownContainer: 5,
  Popup: 3,
  Loader: 1,
  ModalHeader: 1,
  ModalFooter: 1,
  ModalCross: 1,
  ModalBody: 0 };

var priorityStep = 1000;
var getZIndexes = function getZIndexes() {return window.__RetailUiZIndexes || (window.__RetailUiZIndexes = [0]);};
var getIndexPriority = function getIndexPriority(zIndex) {return Math.trunc(zIndex / priorityStep);};
var getMaxAllowedValue = function getMaxAllowedValue(priority) {return (priority + 1) * priorityStep - 1;};
exports.upperBorder = getMaxAllowedValue(Math.max.apply(Math, Object.values(componentPriorities)));
function incrementZIndex(priority, delta) {
  if (typeof priority === 'string') {
    priority = componentPriorities[priority];
  }
  var maxAllowedValue = getMaxAllowedValue(priority);
  var zIndexes = getZIndexes();
  var prevIndexId = zIndexes.length - 1;
  while (zIndexes[prevIndexId] > maxAllowedValue && prevIndexId > 0) {
    prevIndexId--;
  }
  var zIndex = getIndexPriority(zIndexes[prevIndexId]) === priority ?
  Math.min(zIndexes[prevIndexId] + delta, maxAllowedValue) :
  priority * priorityStep;
  zIndexes.push(zIndex);
  zIndexes.sort(function (a, b) {return a - b;});
  return zIndex;
}
exports.incrementZIndex = incrementZIndex;
function removeZIndex(zIndex) {
  var zIndexes = getZIndexes();
  var i = zIndexes.indexOf(zIndex);
  if (i !== -1) {
    zIndexes.splice(i, 1);
  }
}
exports.removeZIndex = removeZIndex;
