// @skbkontur/react-ui/components/CurrencyInput/CurrencyInputKeyboardActions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Keyboard = tslib_1.__importStar(require("../../lib/events/keyboard/identifiers"));
var extractKeyboardAction_1 = require("../internal/extractKeyboardAction");
exports.CURRENCY_INPUT_ACTIONS = {
  Unknown: 0,
  Ignore: 1,
  MoveCursorLeft: 11,
  MoveCursorRight: 12,
  Home: 13,
  End: 14,
  ExtendSelectionLeft: 21,
  ExtendSelectionRight: 22,
  FullSelection: 23,
  ExtendSelectionToStart: 24,
  ExtendSelectionToEnd: 25,
  Backspace: 31,
  Delete: 32,
  Submit: 33 };

exports.extractAction = new extractKeyboardAction_1.KeyboardActionExctracterBuilder().
add(exports.CURRENCY_INPUT_ACTIONS.Submit, Keyboard.isKeyEnter).
add(exports.CURRENCY_INPUT_ACTIONS.ExtendSelectionLeft, Keyboard.isModShift(Keyboard.isKeyArrowLeft)).
add(exports.CURRENCY_INPUT_ACTIONS.ExtendSelectionRight, Keyboard.isModShift(Keyboard.isKeyArrowRight)).
add(exports.CURRENCY_INPUT_ACTIONS.FullSelection, Keyboard.isShortcutSelectAll).
add(exports.CURRENCY_INPUT_ACTIONS.ExtendSelectionToStart, Keyboard.isModShift(Keyboard.isKeyHome)).
add(exports.CURRENCY_INPUT_ACTIONS.ExtendSelectionToEnd, Keyboard.isModShift(Keyboard.isKeyEnd)).
add(exports.CURRENCY_INPUT_ACTIONS.MoveCursorLeft, Keyboard.isKeyArrowLeft).
add(exports.CURRENCY_INPUT_ACTIONS.MoveCursorRight, Keyboard.isKeyArrowRight).
add(exports.CURRENCY_INPUT_ACTIONS.Home, Keyboard.isUnmodified(Keyboard.isKeyHome)).
add(exports.CURRENCY_INPUT_ACTIONS.End, Keyboard.isUnmodified(Keyboard.isKeyEnd)).
add(exports.CURRENCY_INPUT_ACTIONS.Backspace, Keyboard.isKeyBackspace).
add(exports.CURRENCY_INPUT_ACTIONS.Delete, Keyboard.isKeyDelete).
add(exports.CURRENCY_INPUT_ACTIONS.Ignore, function (e) {return Keyboard.isModified()(e) || Keyboard.isKeyTab(e);}).
build(exports.CURRENCY_INPUT_ACTIONS.Unknown);
