// @skbkontur/react-ui/lib/theming/Emotion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var _a;
var create_emotion_1 = tslib_1.__importDefault(require("create-emotion"));
var stylis_plugin_extra_scope_1 = tslib_1.__importDefault(require("stylis-plugin-extra-scope"));
var Upgrades_1 = tslib_1.__importDefault(require("../Upgrades"));
var variables_module_less_1 = tslib_1.__importDefault(require("../../components/variables.module.css"));
var PREFIX = 'react-ui';
// NOTE: for ones who overrides specificityLevel at custom less
var specificityLevel = parseInt(variables_module_less_1["default"].specificityLevel, 10) || 0;
if (specificityLevel) {
  Upgrades_1["default"].setSpecificityLevel(specificityLevel);
}
var scope = new Array(Upgrades_1["default"].getSpecificityLevel()).fill("." + PREFIX).join('');
exports.flush = (_a = create_emotion_1["default"]({
  key: PREFIX,
  stylisPlugins: scope ? [stylis_plugin_extra_scope_1["default"](scope)] : undefined }),
_a.flush), exports.hydrate = _a.hydrate, exports.cx = _a.cx, exports.merge = _a.merge, exports.getRegisteredStyles = _a.getRegisteredStyles, exports.injectGlobal = _a.injectGlobal, exports.keyframes = _a.keyframes, exports.css = _a.css, exports.sheet = _a.sheet, exports.cache = _a.cache;
function prefixer(classes) {
  return Object.entries(classes).reduce(function (pc, _a) {
    var key = _a[0],value = _a[1];
    var _b;
    return tslib_1.__assign({}, pc, (_b = {}, _b[key] = PREFIX + "-" + value + "-" + Math.floor(Math.random() * 1e8).toString(16), _b));
  }, {});
}
exports.prefixer = prefixer;
