// @skbkontur/react-ui/lib/date/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var types_1 = require("./types");
exports.MIN_YEAR = 0;
exports.MAX_YEAR = 9999;
exports.MIN_MONTH = 1;
exports.MAX_MONTH = 12;
exports.MIN_DATE = 1;
exports.MAX_DATE = 31;
exports.MIN_FULLDATE = '01.01.1900';
exports.MAX_FULLDATE = '31.12.2099';
exports.LENGTH_YEAR = 4;
exports.LENGTH_MONTH = 2;
exports.LENGTH_DATE = 2;
exports.LENGTH_SEPARATOR = 1;
exports.CHAR_PAD = '0';
exports.CHAR_MASK = '_';
exports.SEPARATOR = "(?:\\.|\\/|\\-|\\s|\\,)";
exports.RE_ORDER_MDY = new RegExp("^(\\d{1," + exports.LENGTH_MONTH + "})?" + exports.SEPARATOR + "?(\\d{1," + exports.LENGTH_DATE + "})?" + exports.SEPARATOR + "?(\\d{1," + exports.LENGTH_YEAR + "})?$");
exports.RE_ORDER_DMY = new RegExp("^(\\d{1," + exports.LENGTH_DATE + "})?" + exports.SEPARATOR + "?(\\d{1," + exports.LENGTH_MONTH + "})?" + exports.SEPARATOR + "?(\\d{1," + exports.LENGTH_YEAR + "})?$");
exports.RE_ORDER_YMD = new RegExp("^(\\d{1," + exports.LENGTH_YEAR + "})?" + exports.SEPARATOR + "?(\\d{1," + exports.LENGTH_MONTH + "})?" + exports.SEPARATOR + "?(\\d{1," + exports.LENGTH_DATE + "})?$");
exports.emptyDateComponents = {
  year: null,
  month: null,
  date: null };

exports.defaultDateComponentsOrder = types_1.InternalDateOrder.DMY;
exports.defaultDateComponentsSeparator = types_1.InternalDateSeparator.Dot;
exports.defaultFirstDayWeek = types_1.InternalDateFirstDayWeek.Monday;
