// @skbkontur/react-ui/components/SidePage/SidePageFooter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var SidePageContext_1 = require("./SidePageContext");
var SidePage_module_less_1 = tslib_1.__importDefault(require("./SidePage.module.css"));
var LayoutEvents_1 = tslib_1.__importDefault(require("../../lib/LayoutEvents"));
var utils_1 = require("../../lib/utils");
var Emotion_1 = require("../../lib/theming/Emotion");
var SidePage_styles_1 = tslib_1.__importDefault(require("./SidePage.styles"));
var ThemeConsumer_1 = require("../ThemeConsumer");
/**
                                                    * Футер модального окна.
                                                    */
var SidePageFooter = /** @class */function (_super) {
  tslib_1.__extends(SidePageFooter, _super);
  function SidePageFooter() {
    var _this = _super !== null && _super.apply(this, arguments) || this;
    _this.state = {
      fixed: false };

    _this.content = null;
    _this.wrapper = null;
    _this.layoutSub = null;
    _this.update = function () {
      _this.setProperStyles();
    };
    _this.refContent = function (node) {
      _this.content = node;
    };
    _this.refWrapper = function (node) {
      _this.wrapper = node;
    };
    _this.setProperStyles = function () {
      if (_this.wrapper && _this.content) {
        var wrapperRect = _this.wrapper.getBoundingClientRect();
        var contentRect = _this.content.getBoundingClientRect();
        var fixed = wrapperRect.top > contentRect.top;
        _this.setState({ fixed: fixed });
      }
    };
    return _this;
  }
  SidePageFooter.prototype.componentDidMount = function () {
    var context = this.props.context;
    if (context) {
      context.footerRef(this);
    }
    this.update();
    this.layoutSub = LayoutEvents_1["default"].addListener(this.update);
  };
  SidePageFooter.prototype.componentWillUnmount = function () {
    var context = this.props.context;
    if (context) {
      context.footerRef(null);
    }
    if (this.layoutSub) {
      this.layoutSub.remove();
    }
  };
  SidePageFooter.prototype.render = function () {
    var _this = this;
    return React.createElement(ThemeConsumer_1.ThemeConsumer, null, function (theme) {
      _this.theme = theme;
      return _this.renderMain();
    });
  };
  SidePageFooter.prototype.renderMain = function () {
    var _this = this;
    return React.createElement("div", { style: { height: this.getContentHeight() }, ref: this.refWrapper },
    React.createElement(SidePageContext_1.SidePageContext.Consumer, null, function (_a) {
      var getWidth = _a.getWidth;
      var _b;
      return React.createElement("div", { className: SidePage_module_less_1["default"].footer, style: {
          width: getWidth() } },

      React.createElement("div", { className: Emotion_1.cx(SidePage_module_less_1["default"].footerContent, (_b = {},
        _b[SidePage_module_less_1["default"].panel] = !!_this.props.panel,
        _b[SidePage_styles_1["default"].panel(_this.theme)] = !!_this.props.panel,
        _b[SidePage_module_less_1["default"].fixed] = _this.state.fixed,
        _b[SidePage_styles_1["default"].fixed(_this.theme)] = _this.state.fixed,
        _b)), ref: _this.refContent }, _this.props.children));
    }));
  };
  SidePageFooter.prototype.getContentHeight = function () {
    if (!this.content) {
      return 'auto';
    }
    return this.content.getBoundingClientRect().height;
  };
  SidePageFooter.__KONTUR_REACT_UI__ = 'SidePageFooter';
  return SidePageFooter;
}(React.Component);
exports.SidePageFooter = SidePageFooter;
exports.SidePageFooterWithContext = utils_1.withContext(SidePageContext_1.SidePageContext.Consumer)(SidePageFooter);
exports["default"] = exports.SidePageFooterWithContext;
