// @skbkontur/react-ui/components/Spinner/Spinner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var PropTypes = tslib_1.__importStar(require("prop-types"));
var decorators_1 = require("../LocaleProvider/decorators");
var locale_1 = require("./locale");
var settings_1 = require("./settings");
var Spinner_module_less_1 = tslib_1.__importDefault(require("./Spinner.module.css"));
var SpinnerFallback_1 = tslib_1.__importDefault(require("./SpinnerFallback"));
var Spinner_styles_1 = tslib_1.__importDefault(require("./Spinner.styles"));
var Emotion_1 = require("../../lib/theming/Emotion");
var ThemeConsumer_1 = require("../ThemeConsumer");
exports.SpinnerConfig = {
  hasSvgAnimationSupport: settings_1.svgAnimateSupport() };

var CLOUD_SVG_PATH = "M32.0297086,9.1495774 L31.5978628,8.5870774 C29.3570968,\n      5.67148577 25.9305165,4 22.1999329,4 C17.3429265,\n      4 12.9026663,7.04270463 11.154144,11.5717304 L10.901479,\n      12.2253114 L10.2421341,12.4725311 C6.50853057,13.8727758 4,\n      17.4719751 4,21.428492 C4,26.7061833 8.32047079,\n      31 13.6314689,31 L32.0297086,31 C38.078569,31 43,26.1036477 43,\n      20.0862989 C43,14.3602091 38.493302,9.5769573 32.7403918,\n      9.19661922 L32.0297086,9.1495774 Z";
/**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        * DRAFT - инлайн-лоадер
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        */
var Spinner = /** @class */function (_super) {
  tslib_1.__extends(Spinner, _super);
  function Spinner() {
    var _this = _super !== null && _super.apply(this, arguments) || this;
    _this._renderCloud = function (type) {
      var params = settings_1.sizeMaps[type];
      var bgClassName = Spinner_styles_1["default"].cloudBg(_this.theme);
      var strokeClassName = Emotion_1.cx(Spinner_module_less_1["default"].cloudStroke, _this.props.dimmed ? Spinner_styles_1["default"].cloudStrokeDimmed(_this.theme) : Spinner_styles_1["default"].cloudStroke(_this.theme));
      return React.createElement("svg", { className: Spinner_module_less_1["default"].cloud, width: params.width, height: params.height, viewBox: params.viewBox },
      React.createElement("path", { d: CLOUD_SVG_PATH, strokeWidth: params.strokeWidth, className: bgClassName }),
      React.createElement("path", { d: CLOUD_SVG_PATH, strokeWidth: params.strokeWidth, className: strokeClassName }));
    };
    _this._renderCircle = function (type) {
      var params = settings_1.sizeMaps[type];
      var theme = _this.theme;
      var strokeClassName = _this.props.dimmed ? Spinner_styles_1["default"].circleStrokeDimmed(theme) : Spinner_styles_1["default"].circleStroke(theme);
      return React.createElement("svg", { className: Emotion_1.cx(Spinner_module_less_1["default"].circle, Spinner_styles_1["default"].circle(theme)), width: params.width, height: params.height },
      React.createElement("circle", { cx: "8", cy: "8", r: "6", strokeWidth: params.strokeWidth, className: strokeClassName }));
    };
    _this._renderSpinner = function (type) {
      if (type === settings_1.types.mini) {
        return _this._renderCircle(type);
      }
      return _this._renderCloud(type);
    };
    _this._renderCaption = function (type, caption) {
      var captionClassName = Emotion_1.cx(Spinner_module_less_1["default"].caption, Spinner_styles_1["default"].caption(_this.theme), type === settings_1.types.mini ? Spinner_module_less_1["default"].captionRight : Spinner_module_less_1["default"].captionBottom);
      return React.createElement("span", { className: captionClassName }, caption);
    };
    return _this;
  }
  Spinner_1 = Spinner;
  Spinner.prototype.render = function () {
    var _this = this;
    return React.createElement(ThemeConsumer_1.ThemeConsumer, null, function (theme) {
      _this.theme = theme;
      return _this.renderMain();
    });
  };
  Spinner.prototype.renderMain = function () {
    var _a = this.props,type = _a.type,_b = _a.caption,caption = _b === void 0 ? this.locale.loading : _b,dimmed = _a.dimmed;
    var verifiedType = settings_1.sizeMaps[type] ? type : Spinner_1.defaultProps.type;
    return React.createElement("div", { className: Spinner_module_less_1["default"].spinner },
    React.createElement("span", { className: Spinner_module_less_1["default"].inner },
    exports.SpinnerConfig.hasSvgAnimationSupport && this._renderSpinner(verifiedType),
    !exports.SpinnerConfig.hasSvgAnimationSupport && React.createElement(SpinnerFallback_1["default"], { type: verifiedType, dimmed: dimmed })),
    caption && this._renderCaption(verifiedType, caption));
  };
  var Spinner_1;
  Spinner.__KONTUR_REACT_UI__ = 'Spinner';
  Spinner.propTypes = {
    /**
                         * Текст рядом с мини-лоадером.
                         *
                         * 'Загрузка' - значение по-умолчанию
                         */
    caption: PropTypes.node,
    dimmed: PropTypes.bool,
    /**
                             * Тип спиннера: mini, normal, big
                             *
                             * Значение по-умолчанию - normal
                             *
                             * Spinner.types - все доступные типы
                             */
    type: PropTypes.oneOf(Object.keys(settings_1.types)) };

  Spinner.defaultProps = {
    type: settings_1.types.normal };

  Spinner = Spinner_1 = tslib_1.__decorate([
  decorators_1.locale('Spinner', locale_1.SpinnerLocaleHelper)],
  Spinner);
  return Spinner;
}(React.Component);
Spinner.Types = settings_1.types;
exports["default"] = Spinner;
