// @skbkontur/react-ui/components/Calendar/Animation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var raf_1 = tslib_1.__importStar(require("raf"));
var stepper_1 = require("./stepper");
var noop = function noop() {
  var args = [];
  for (var _i = 0; _i < arguments.length; _i++) {
    args[_i] = arguments[_i];
  }
  /* do nothing */
};
exports.Animation = function () {
  var animating = false;
  var currentPosition = 0;
  var currentVelocity = 0;
  var rafId = 0;
  var target = 0;
  var deltaHandler = noop;
  var animationEndHandler = noop;
  function reset() {
    currentVelocity = 0;
    currentPosition = 0;
    target = 0;
    animating = false;
    deltaHandler = noop;
    animationEndHandler = noop;
  }
  function inProgress() {
    return animating;
  }
  function cancel() {
    raf_1.cancel(rafId);
    reset();
  }
  function finish() {
    animationEndHandler();
    cancel();
  }
  function animate(amount, onDelta, onEnd) {
    if (onEnd === void 0) {onEnd = noop;}
    target += amount;
    deltaHandler = onDelta;
    animationEndHandler = onEnd;
    var animateInternal = function animateInternal() {
      animating = true;
      var _a = stepper_1.stepper(currentPosition, currentVelocity, target),nextPosition = _a[0],nextVelocity = _a[1];
      var delta = nextPosition - currentPosition;
      deltaHandler(delta);
      if (nextPosition === target && nextVelocity === 0) {
        animationEndHandler();
        reset();
        return;
      }
      currentPosition = nextPosition;
      currentVelocity = nextVelocity;
      rafId = raf_1["default"](animateInternal);
    };
    if (!inProgress()) {
      animateInternal();
    }
  }
  return {
    animate: animate,
    inProgress: inProgress,
    cancel: cancel,
    finish: finish };

};
