// @skbkontur/react-ui/components/Calendar/CalendarDateShape.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isEqual = function (a, b) {
  return !a && !b || a && b && a.year === b.year && a.month === b.month && a.date === b.date;
};
exports.comparator = function (a, b) {
  if (a.year < b.year) {
    return -1;
  } else
  if (a.year > b.year) {
    return 1;
  } else
  if (a.month < b.month) {
    return -1;
  } else
  if (a.month > b.month) {
    return 1;
  } else
  if (a.date < b.date) {
    return -1;
  } else
  if (a.date > b.date) {
    return 1;
  }
  return 0;
};
exports.create = function (date, month, year) {return { date: date, month: month, year: year };};
exports.isLess = function (left, right) {return exports.comparator(left, right) === -1;};
exports.isLessOrEqual = function (left, right) {
  return exports.isLess(left, right) || exports.isEqual(left, right);
};
exports.isGreater = function (left, right) {return !exports.isLessOrEqual(left, right);};
exports.isGreaterOrEqual = function (left, right) {return !exports.isLess(left, right);};
exports.isBetween = function (date, left, right) {
  if (left && exports.isLess(date, left)) {
    return false;
  }
  if (right && exports.isGreater(date, right)) {
    return false;
  }
  return true;
};
