// @skbkontur/react-ui/components/Fias/FiasSearch/FiasSearch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var locale_1 = require("../locale");
var FiasComboBox_1 = require("../Form/FiasComboBox");
var Address_1 = require("../models/Address");
var types_1 = require("../types");
var decorators_1 = require("../../LocaleProvider/decorators");
var filterProps_1 = tslib_1.__importDefault(require("../../filterProps"));
var COMBOBOX_PASS_PROPS = {
  limit: true,
  align: true,
  searchOnFocus: true,
  drawArrow: true,
  autoFocus: true,
  borderless: true,
  disablePortal: true,
  disabled: true,
  error: true,
  maxLength: true,
  menuAlign: true,
  onBlur: true,
  onFocus: true,
  onInputChange: true,
  placeholder: true,
  size: true,
  warning: true,
  width: true,
  maxMenuHeight: true,
  onMouseEnter: true,
  onMouseOver: true,
  onMouseLeave: true };

var FiasSearch = /** @class */function (_super) {
  tslib_1.__extends(FiasSearch, _super);
  function FiasSearch() {
    var _this = _super !== null && _super.apply(this, arguments) || this;
    _this.renderItem = function (address) {
      return address.getText();
    };
    _this.renderValue = function (address) {
      return address.getText(types_1.Fields.room);
    };
    _this.renderNotFound = function () {
      return _this.locale.searchNotFound;
    };
    _this.valueToString = function (address) {
      return address.getText(types_1.Fields.room);
    };
    _this.onChange = function (event, address) {
      var onChange = _this.props.onChange;
      if (onChange) {
        onChange(event, address);
      }
    };
    _this.onUnexpectedInput = function (query) {
      if (!query) {
        return new Address_1.Address();
      }
    };
    _this.getItems = function (searchText) {return tslib_1.__awaiter(_this, void 0, void 0, function () {
        var _a, api, limit, options;
        return tslib_1.__generator(this, function (_b) {
          _a = this.props, api = _a.api, limit = _a.limit;
          options = {
            searchText: searchText,
            fullAddress: true,
            directParent: false,
            limit: (limit || FiasSearch_1.defaultProps.limit) + 1 };

          return [2 /*return*/, api.search(options).then(function (result) {
            var success = result.success,data = result.data,error = result.error;
            return success && data ?
            Promise.resolve(data.map(function (item) {return Address_1.Address.createFromResponse(item);})) :
            Promise.reject(error);
          })];
        });
      });};
    return _this;
  }
  FiasSearch_1 = FiasSearch;
  FiasSearch.prototype.render = function () {
    var restComboBoxProps = filterProps_1["default"](this.props, COMBOBOX_PASS_PROPS);
    return React.createElement(FiasComboBox_1.FiasComboBox, tslib_1.__assign({ getItems: this.getItems, value: this.props.address, renderItem: this.renderItem, renderValue: this.renderValue, valueToString: this.valueToString, onChange: this.onChange, onUnexpectedInput: this.onUnexpectedInput, renderNotFound: this.renderNotFound }, restComboBoxProps));
  };
  var FiasSearch_1;
  FiasSearch.__KONTUR_REACT_UI__ = 'FiasSearch';
  FiasSearch.defaultProps = {
    width: '100%',
    limit: 5,
    drawArrow: false,
    searchOnFocus: false };

  FiasSearch = FiasSearch_1 = tslib_1.__decorate([
  decorators_1.locale('Fias', locale_1.FiasLocaleHelper)],
  FiasSearch);
  return FiasSearch;
}(React.Component);
exports.FiasSearch = FiasSearch;
exports["default"] = FiasSearch;
