// @skbkontur/react-ui/components/Paging/Paging.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Emotion_1 = require("../../lib/theming/Emotion");
var Paging_module_less_1 = tslib_1.__importDefault(require("./Paging.module.css"));
var jsStyles = {
  dots: function dots(t) {
    return Emotion_1.css(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      color: ", ";\n    "], ["\n      color: ", ";\n    "])), t.pagingDotsColor);
  },
  forwardLink: function forwardLink(t) {
    return Emotion_1.css(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n      color: ", ";\n    "], ["\n      color: ", ";\n    "])), t.pagingForwardLinkColor);
  },
  disabled: function disabled(t) {
    return Emotion_1.css(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n      .", "& {\n        color: ", ";\n      }\n    "], ["\n      .", "& {\n        color: ", ";\n      }\n    "])), Paging_module_less_1["default"].forwardLink, t.pagingForwardLinkDisabledColor);
  },
  pageLink: function pageLink(t) {
    return Emotion_1.css(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n      .", " & {\n        color: ", ";\n      }\n      .", " &:not(.", "):hover {\n        background: ", ";\n      }\n    "], ["\n      .", " & {\n        color: ", ";\n      }\n      .", " &:not(.", "):hover {\n        background: ", ";\n      }\n    "])), Paging_module_less_1["default"].pageLinkWrapper, t.pagingForwardLinkColor, Paging_module_less_1["default"].pageLinkWrapper, Paging_module_less_1["default"].active, t.pagingPageLinkHoverBg);
  },
  active: function active(t) {
    return Emotion_1.css(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n      .", " .", "& {\n        background: ", ";\n        color: ", ";\n      }\n    "], ["\n      .", " .", "& {\n        background: ", ";\n        color: ", ";\n      }\n    "])), Paging_module_less_1["default"].pageLinkWrapper, Paging_module_less_1["default"].pageLink, t.pagingPageLinkActiveBg, t.pagingPageLinkActiveColor);
  },
  focused: function focused(t) {
    return Emotion_1.css(templateObject_6 || (templateObject_6 = tslib_1.__makeTemplateObject(["\n      .", " .", "& {\n        border: solid 2px ", ";\n      }\n    "], ["\n      .", " .", "& {\n        border: solid 2px ", ";\n      }\n    "])), Paging_module_less_1["default"].pageLinkWrapper, Paging_module_less_1["default"].pageLink, t.borderColorFocus);
  },
  pageLinkHint: function pageLinkHint(t) {
    return Emotion_1.css(templateObject_7 || (templateObject_7 = tslib_1.__makeTemplateObject(["\n      .", " & {\n        color: ", ";\n      }\n    "], ["\n      .", " & {\n        color: ", ";\n      }\n    "])), Paging_module_less_1["default"].pageLinkWrapper, t.pagingPageLinkHintColor);
  } };

exports["default"] = jsStyles;
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7;
