// @skbkontur/react-ui/components/TokenInput/TextWidthHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var RenderContainer_1 = require("../RenderContainer/RenderContainer");
var styles = {
  position: 'absolute',
  left: -1000,
  top: -1000,
  fontSize: 14 };

var TextWidthHelper = /** @class */function (_super) {
  tslib_1.__extends(TextWidthHelper, _super);
  function TextWidthHelper() {
    var _this = _super !== null && _super.apply(this, arguments) || this;
    _this.element = null;
    _this.elementRef = function (node) {return _this.element = node;};
    return _this;
  }
  TextWidthHelper.prototype.render = function () {
    return React.createElement(RenderContainer_1.RenderContainer, null,
    React.createElement("div", { style: styles, ref: this.elementRef }, this.props.text));
  };
  TextWidthHelper.prototype.getTextWidth = function () {
    return this.element.getBoundingClientRect().width;
  };
  return TextWidthHelper;
}(React.Component);
exports["default"] = TextWidthHelper;
