// @skbkontur/react-ui/lib/theming/themes/FlatTheme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var variables_flat_module_less_1 = tslib_1.__importDefault(require("../../../components/variables.flat.module.css"));
var ThemeHelpers_1 = require("../ThemeHelpers");
var FLAT_THEME = ThemeHelpers_1.defineInternalTheme(variables_flat_module_less_1["default"], {
  btnFontSizeMedium: {
    get: function get() {
      return '16px';
    } },

  btnCheckedShadowArrow: {
    get: function get() {
      return "0 0 0 1px " + this.btnCheckedBg;
    } },

  btnCheckedShadowArrowLeft: {
    get: function get() {
      return "0 0 0 1px " + this.btnCheckedBg;
    } },

  btnFocusBorder: {
    get: function get() {
      return "1px solid " + this.borderColorFocus;
    } },

  btnCheckedHoverBorderColor: {
    get: function get() {
      return this.btnCheckedBg;
    } },

  btnDefaultShadowArrowLeft: {
    get: function get() {
      return this.btnDefaultShadowArrow;
    } },

  btnDefaultCheckedShadowArrow: {
    get: function get() {
      return "1px -1px 0 0 rgba(0, 0, 0, 0.15), 0 0 0 1px " + this.btnCheckedBg;
    } },

  btnDefaultHoverShadowArrow: {
    get: function get() {
      return "1px -1px 0 0 " + this.btnDefaultHoverBorderColor + ", 1px -1px 0 0 " + this.btnDefaultHoverBg + ";";
    } },

  btnDefaultHoverShadowArrowLeft: {
    get: function get() {
      return "1px -1px 0 0 " + this.btnDefaultHoverBorderColor + ", 1px -1px 0 0 " + this.btnDefaultHoverBg + ";";
    } },

  btnDefaultActiveShadowArrow: {
    get: function get() {
      return "1px -1px 0 0 " + this.btnDefaultActiveBorderColor + ", 1px -1px 0 0 " + this.btnDefaultActiveBg + ";";
    } },

  btnDefaultActiveShadowArrowLeft: {
    get: function get() {
      return "1px -1px 0 0 " + this.btnDefaultActiveBorderColor + ", 1px -1px 0 0 " + this.btnDefaultActiveBg + ";";
    } },

  btnPrimaryShadowArrow: {
    get: function get() {
      return "0 0 0 1px " + this.btnPrimaryDisabledBg + ";";
    } },

  btnPrimaryShadowArrowLeft: {
    get: function get() {
      return "0 0 0 1px " + this.btnPrimaryDisabledBg + ";";
    } },

  btnPrimaryHoverShadowArrow: {
    get: function get() {
      return "0 0 0 1px " + this.btnPrimaryHoverBg + ";";
    } },

  btnPrimaryHoverShadowArrowLeft: {
    get: function get() {
      return "0 0 0 1px " + this.btnPrimaryHoverBg + ";";
    } },

  btnPrimaryActiveShadowArrow: {
    get: function get() {
      return "0 0 0 1px " + this.btnPrimaryActiveBg + ";";
    } },

  btnPrimaryActiveShadowArrowLeft: {
    get: function get() {
      return "0 0 0 1px " + this.btnPrimaryActiveBg + ";";
    } },

  btnSuccessShadowArrow: {
    get: function get() {
      return "0 0 0 1px " + this.btnSuccessDisabledBg + ";";
    } },

  btnSuccessShadowArrowLeft: {
    get: function get() {
      return "0 0 0 1px " + this.btnSuccessDisabledBg + ";";
    } },

  btnSuccessHoverShadowArrow: {
    get: function get() {
      return "0 0 0 1px " + this.btnSuccessHoverBg + ";";
    } },

  btnSuccessHoverShadowArrowLeft: {
    get: function get() {
      return "0 0 0 1px " + this.btnSuccessHoverBg + ";";
    } },

  btnSuccessActiveShadowArrow: {
    get: function get() {
      return "0 0 0 1px " + this.btnSuccessActiveBg + ";";
    } },

  btnSuccessActiveShadowArrowLeft: {
    get: function get() {
      return "0 0 0 1px " + this.btnSuccessActiveBg + ";";
    } },

  btnDangerShadowArrow: {
    get: function get() {
      return "0 0 0 1px " + this.btnDangerDisabledBg + ";";
    } },

  btnDangerShadowArrowLeft: {
    get: function get() {
      return "0 0 0 1px " + this.btnDangerDisabledBg + ";";
    } },

  btnDangerHoverShadowArrow: {
    get: function get() {
      return "0 0 0 1px " + this.btnDangerHoverBg + ";";
    } },

  btnDangerHoverShadowArrowLeft: {
    get: function get() {
      return "0 0 0 1px " + this.btnDangerHoverBg + ";";
    } },

  btnDangerActiveShadowArrow: {
    get: function get() {
      return "0 0 0 1px " + this.btnDangerActiveBg + ";";
    } },

  btnDangerActiveShadowArrowLeft: {
    get: function get() {
      return "0 0 0 1px " + this.btnDangerActiveBg + ";";
    } },

  btnPayShadowArrow: {
    get: function get() {
      return "0 0 0 1px " + this.btnPayDisabledBg + ";";
    } },

  btnPayShadowArrowLeft: {
    get: function get() {
      return "0 0 0 1px " + this.btnPayDisabledBg + ";";
    } },

  btnPayHoverShadowArrow: {
    get: function get() {
      return "0 0 0 1px " + this.btnPayHoverBg + ";";
    } },

  btnPayHoverShadowArrowLeft: {
    get: function get() {
      return "0 0 0 1px " + this.btnPayHoverBg + ";";
    } },

  btnPayActiveShadowArrow: {
    get: function get() {
      return "0 0 0 1px " + this.btnPayActiveBg + ";";
    } },

  btnPayActiveShadowArrowLeft: {
    get: function get() {
      return "0 0 0 1px " + this.btnPayActiveBg + ";";
    } },

  inputDisabledBorderColor: {
    get: function get() {
      return this.inputDisabledBg;
    } },

  chbBorder: {
    get: function get() {
      return this.btnDefaultBorder;
    } },

  chbBorderColorFocus: {
    get: function get() {
      return this.borderColorFocus;
    } },

  chbHoverBg: {
    get: function get() {
      return this.btnDefaultHoverBg;
    } },

  chbActiveBg: {
    get: function get() {
      return this.btnDefaultActiveBg;
    } },

  chbCheckedBg: {
    get: function get() {
      return this.borderColorFocus;
    } },

  chbBoxIndeterminateBg: {
    get: function get() {
      return this.chbCheckedBg;
    } },

  radioHoverBg: {
    get: function get() {
      return this.chbHoverBg;
    } },

  radioActiveBg: {
    get: function get() {
      return this.chbActiveBg;
    } },

  radioFocusShadow: {
    get: function get() {
      return "inset 0 0 0 1px " + this.outlineColorFocus + ";";
    } },

  radioCheckedBgColor: {
    get: function get() {
      return this.chbCheckedBg;
    } },

  textareaDisabledBorderColor: {
    get: function get() {
      return this.textareaDisabledBg;
    } } });


exports["default"] = FLAT_THEME;
