// @skbkontur/react-ui/components/IgnoreLayerClick/IgnoreLayerClick.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var react_dom_1 = require("react-dom");
var IgnoreLayerClick = /** @class */function (_super) {
  tslib_1.__extends(IgnoreLayerClick, _super);
  function IgnoreLayerClick() {
    return _super !== null && _super.apply(this, arguments) || this;
  }
  IgnoreLayerClick.prototype.render = function () {
    var child = React.Children.only(this.props.children);
    return this.props.active ? React.createElement(IgnoreLayerClickWrapper, null, child) : child;
  };
  return IgnoreLayerClick;
}(React.Component);
exports["default"] = IgnoreLayerClick;
// NOTE Используется только в команде Контур.Бухгалтерия
var IgnoreLayerClickWrapper = /** @class */function (_super) {
  tslib_1.__extends(IgnoreLayerClickWrapper, _super);
  function IgnoreLayerClickWrapper() {
    var _this = _super !== null && _super.apply(this, arguments) || this;
    _this.element = null;
    _this.handleMouseDown = function (event) {
      event.stopPropagation();
    };
    return _this;
  }
  IgnoreLayerClickWrapper.prototype.componentDidMount = function () {
    var element = react_dom_1.findDOMNode(this);
    if (element && element instanceof Element) {
      element.addEventListener('mousedown', this.handleMouseDown);
      this.element = element;
    }
  };
  IgnoreLayerClickWrapper.prototype.componentWillUnmount = function () {
    if (this.element) {
      this.element.removeEventListener('mousedown', this.handleMouseDown);
      this.element = null;
    }
  };
  IgnoreLayerClickWrapper.prototype.render = function () {
    return React.Children.only(this.props.children);
  };
  IgnoreLayerClickWrapper.__KONTUR_REACT_UI__ = 'IgnoreLayerClick';
  return IgnoreLayerClickWrapper;
}(React.Component);
