// @skbkontur/react-ui/components/DateSelect/DateSelect.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Emotion_1 = require("../../lib/theming/Emotion");
var jsStyles = {
  root: function root(t) {
    return Emotion_1.css(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      color: ", ";\n\n      &:hover {\n        color: ", ";\n      }\n    "], ["\n      color: ", ";\n\n      &:hover {\n        color: ", ";\n      }\n    "])), t.textColorDefault, t.linkColor);
  },
  arrow: function arrow(t) {
    return Emotion_1.css(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n      color: ", ";\n    "], ["\n      color: ", ";\n    "])), t.textColorDisabled);
  },
  menuHolder: function menuHolder(t) {
    return Emotion_1.css(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n      background: ", ";\n      box-shadow: ", ";\n      color: ", ";\n    "], ["\n      background: ", ";\n      box-shadow: ", ";\n      color: ", ";\n    "])), t.dateSelectMenuBg, t.popupBoxShadow, t.textColorDefault);
  },
  menuItem: function menuItem(t) {
    return Emotion_1.css(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n      color: ", ";\n    "], ["\n      color: ", ";\n    "])), t.textColorDefault);
  },
  menuItemSelected: function menuItemSelected(t) {
    return Emotion_1.css(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n      background: ", ";\n      color: ", ";\n    "], ["\n      background: ", ";\n      color: ", ";\n    "])), t.dateSelectMenuItemBgSelected, t.dateSelectMenuItemFontSelected);
  },
  menuItemActive: function menuItemActive(t) {
    return Emotion_1.css(templateObject_6 || (templateObject_6 = tslib_1.__makeTemplateObject(["\n      background: ", ";\n      color: ", ";\n    "], ["\n      background: ", ";\n      color: ", ";\n    "])), t.dateSelectMenuItemBgActive, t.dateSelectMenuItemFontActive);
  },
  menuItemDisabled: function menuItemDisabled(t) {
    return Emotion_1.css(templateObject_7 || (templateObject_7 = tslib_1.__makeTemplateObject(["\n      color: ", ";\n      background: ", ";\n    "], ["\n      color: ", ";\n      background: ", ";\n    "])), t.textColorDisabled, t.dateSelectMenuItemBgDisabled);
  },
  menuUp: function menuUp(t) {
    return Emotion_1.css(templateObject_8 || (templateObject_8 = tslib_1.__makeTemplateObject(["\n      background: ", ";\n      color: ", ";\n\n      &:hover {\n        background: ", ";\n        color: ", ";\n      }\n    "], ["\n      background: ", ";\n      color: ", ";\n\n      &:hover {\n        background: ", ";\n        color: ", ";\n      }\n    "])), t.dateSelectMenuBg, t.textColorDefault, t.dateSelectMenuItemBgActive, t.textColorInvert);
  },
  menuDown: function menuDown(t) {
    return Emotion_1.css(templateObject_9 || (templateObject_9 = tslib_1.__makeTemplateObject(["\n      background: ", ";\n      color: ", ";\n\n      &:hover {\n        background: ", ";\n        color: ", ";\n      }\n    "], ["\n      background: ", ";\n      color: ", ";\n\n      &:hover {\n        background: ", ";\n        color: ", ";\n      }\n    "])), t.dateSelectMenuBg, t.textColorDefault, t.dateSelectMenuItemBgActive, t.textColorInvert);
  } };

exports["default"] = jsStyles;
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9;
