// @skbkontur/react-ui/components/PasswordInput/PasswordInputFallback.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var Input_1 = tslib_1.__importDefault(require("../Input"));
var PasswordInputFallback = /** @class */function (_super) {
  tslib_1.__extends(PasswordInputFallback, _super);
  function PasswordInputFallback() {
    var _this = _super !== null && _super.apply(this, arguments) || this;
    _this._passwordInput = null;
    _this._textInput = null;
    // sad ie8 hack that move caret position to the end
    _this._handleFocus = function (_a) {
      var target = _a.target;
      var val = target.value;
      target.value = '';
      target.value = val;
    };
    return _this;
  }
  PasswordInputFallback.prototype.componentDidMount = function () {
    if (this._passwordInput) {
      this.props.refInput(this._passwordInput);
    }
  };
  PasswordInputFallback.prototype.componentDidUpdate = function (prevProps) {
    if (prevProps.visible === this.props.visible) {
      return;
    }
    var _a = this.props,visible = _a.visible,refInput = _a.refInput;
    var currentInput = visible ? this._textInput : this._passwordInput;
    if (currentInput) {
      refInput(currentInput);
    }
  };
  PasswordInputFallback.prototype.render = function () {
    var _this = this;
    var visible = this.props.visible;
    return React.createElement("div", null,
    React.createElement("span", { style: { display: !visible ? 'block' : 'none' } },
    React.createElement(Input_1["default"], tslib_1.__assign({ type: "password", onFocus: this._handleFocus, ref: function ref(_ref) {
        _this._passwordInput = _ref;
      } }, this.props))),
    React.createElement("span", { style: { display: visible ? 'block' : 'none' } },
    React.createElement(Input_1["default"], tslib_1.__assign({ type: "text", onFocus: this._handleFocus, ref: function ref(_ref2) {
        _this._textInput = _ref2;
      } }, this.props))));
  };
  PasswordInputFallback.defaultProps = {
    size: 'small' };

  return PasswordInputFallback;
}(React.Component);
exports["default"] = PasswordInputFallback;
