// @skbkontur/react-ui/components/ThemeShowcase/ThemeShowcase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var ThemeShowcase_module_less_1 = tslib_1.__importDefault(require("./ThemeShowcase.module.css"));
var DefaultTheme_1 = tslib_1.__importDefault(require("../../lib/theming/themes/DefaultTheme"));
var FlatTheme_1 = tslib_1.__importDefault(require("../../lib/theming/themes/FlatTheme"));
var ComboBox_1 = tslib_1.__importDefault(require("../ComboBox"));
var Gapped_1 = tslib_1.__importDefault(require("../Gapped"));
var Link_1 = tslib_1.__importDefault(require("../Link"));
var Sticky_1 = tslib_1.__importDefault(require("../Sticky"));
var ColorFunctions_1 = tslib_1.__importDefault(require("../../lib/styles/ColorFunctions"));
var Tooltip_1 = tslib_1.__importDefault(require("../Tooltip"));
var VariablesCollector_1 = require("./ThemeShowcaseHelpers/VariablesCollector");
var Supports_1 = require("../internal/Supports");
var CSS_TOOLTIP_ALLOWED_POSITIONS = ['bottom left', 'top left'];
var EMPTY_ARRAY = [];
var ALL_VARIABLES = Object.keys(DefaultTheme_1["default"]).
concat(Object.keys(FlatTheme_1["default"])).
filter(function (variable, index, all) {return all.indexOf(variable) === index;});
var ThemeShowcase = /** @class */function (_super) {
  tslib_1.__extends(ThemeShowcase, _super);
  function ThemeShowcase() {
    var _this = _super !== null && _super.apply(this, arguments) || this;
    _this.state = {};
    _this.isUnmounting = false;
    _this.variablesDiff = [];
    _this.getItems = function (query) {
      return Promise.resolve(_this.getValues(query));
    };
    _this.handleVariableChange = function (event, item) {
      if (!_this.isUnmounting) {
        _this.setState({ selectedVariable: item });
      }
    };
    _this.handleUnexpectedVariableInput = function (query) {
      var values = _this.getValues(query);
      if (values.length > 0) {
        return values[0];
      } else
      {
        return _this.resetVariable();
      }
    };
    _this.resetVariable = function () {
      if (!_this.isUnmounting) {
        _this.setState({ selectedVariable: undefined });
      }
    };
    return _this;
  }
  ThemeShowcase.prototype.componentWillMount = function () {
    var _this = this;
    if (this.props.isDebugMode) {
      ALL_VARIABLES.forEach(function (variable) {
        var found = VariablesCollector_1.ALL_USED_VARIABLES.includes(variable);
        if (!found) {
          _this.variablesDiff.push(variable);
        }
      });
    }
  };
  ThemeShowcase.prototype.render = function () {
    var _this = this;
    if (!Supports_1.IS_PROXY_SUPPORTED) {
      return React.createElement("div", null,
      "\u0422\u0430\u0431\u043B\u0438\u0446\u0430 \u0438\u0441\u043F\u043E\u043B\u044C\u0437\u043E\u0432\u0430\u043D\u0438\u044F \u043F\u0435\u0440\u0435\u043C\u0435\u043D\u043D\u044B\u0445 \u0434\u043E\u0441\u0442\u0443\u043F\u043D\u0430 \u0442\u043E\u043B\u044C\u043A\u043E \u0432",
      ' ',
      React.createElement(Link_1["default"], { href: 'https://caniuse.com/#feat=proxy', target: '_blank' }, "\u0431\u0440\u0430\u0443\u0437\u0435\u0440\u0430\u0445 \u0441 \u043F\u043E\u0434\u0434\u0435\u0440\u0436\u043A\u043E\u0439 Proxy"),
      ".");
    }
    var selectedVariable = this.state.selectedVariable;
    var descriptionsToRender = selectedVariable ?
    VariablesCollector_1.COMPONENT_DESCRIPTIONS_BY_VARIABLE[selectedVariable.value] || {} :
    VariablesCollector_1.COMPONENT_DESCRIPTIONS;
    var isDebugMode = this.props.isDebugMode;
    var callsCount = isDebugMode ? "(" + VariablesCollector_1.CALLS_COUNT + " \u0432\u044B\u0437\u043E\u0432\u043E\u0432)" : '';
    var executionTime = isDebugMode ? "\u0421\u0433\u0435\u043D\u0435\u0440\u0438\u0440\u043E\u0432\u0430\u043D\u043E \u0437\u0430 " + VariablesCollector_1.EXECUTION_TIME.toFixed(3) + "ms" : '';
    return React.createElement(Gapped_1["default"], { wrap: true, gap: 30, verticalAlign: 'top' },
    React.createElement("div", null,
    React.createElement(Sticky_1["default"], { side: 'top' },
    React.createElement("div", { className: ThemeShowcase_module_less_1["default"].searchBar, "data-perf-info": executionTime + " " + callsCount },
    React.createElement(Gapped_1["default"], { gap: 15 },
    React.createElement(ComboBox_1["default"], { getItems: this.getItems, value: selectedVariable, onChange: this.handleVariableChange, onUnexpectedInput: this.handleUnexpectedVariableInput, placeholder: 'поиск по названию переменной' }),
    !!selectedVariable && React.createElement(Link_1["default"], { onClick: this.resetVariable }, "\u0441\u0431\u0440\u043E\u0441\u0438\u0442\u044C")))),
    Object.keys(descriptionsToRender).
    sort().
    map(function (componentName) {return React.createElement(ComponentShowcase, { key: componentName, name: componentName, description: descriptionsToRender[componentName], isDebugMode: isDebugMode, onVariableSelect: _this.handleVariableChange });})),
    React.createElement(ShowUnusedVariables, { diff: this.variablesDiff }));
  };
  ThemeShowcase.prototype.componentWillUnmount = function () {
    this.isUnmounting = true;
  };
  ThemeShowcase.prototype.getValues = function (query) {
    var lowerCaseQuery = query && query.toLowerCase().trim();
    var allItems = VariablesCollector_1.ALL_USED_VARIABLES;
    if (lowerCaseQuery) {
      allItems = VariablesCollector_1.ALL_USED_VARIABLES.filter(function (usedVariable) {return usedVariable.toLowerCase().startsWith(lowerCaseQuery);});
    }
    return allItems.map(function (usedVariableName) {return {
        value: usedVariableName,
        label: usedVariableName };
    });
  };
  return ThemeShowcase;
}(React.Component);
exports["default"] = ThemeShowcase;
var ComponentShowcase = /** @class */function (_super) {
  tslib_1.__extends(ComponentShowcase, _super);
  function ComponentShowcase() {
    return _super !== null && _super.apply(this, arguments) || this;
  }
  ComponentShowcase.prototype.render = function () {
    var _this = this;
    var _a = this.props,name = _a.name,description = _a.description,onVariableSelect = _a.onVariableSelect,isDebugMode = _a.isDebugMode;
    var elements = Object.keys(description);
    return React.createElement(React.Fragment, null,
    React.createElement(Sticky_1["default"], { side: 'top', offset: parseInt(ThemeShowcase_module_less_1["default"].searchBarHeight, 10) }, function (isSticky) {return React.createElement("h2", { className: ThemeShowcase_module_less_1["default"].heading + " " + (isSticky && ThemeShowcase_module_less_1["default"].headingSticky) }, _this.props.name);}),
    React.createElement("table", { className: ThemeShowcase_module_less_1["default"].table },
    React.createElement("thead", null,
    React.createElement("tr", null,
    React.createElement("th", { style: { width: 170 } }, "ClassName"),
    React.createElement("th", { style: { width: 210 } }, "Variable Name"),
    React.createElement("th", { style: { width: 250 } }, "Default Value"),
    React.createElement("th", { style: { width: 250 } }, "Flat Value"))),
    React.createElement("tbody", null, elements.map(function (el) {return React.createElement(ComponentShowcaseRow, { key: name + "_" + el, element: el, row: description[el], onVariableSelect: onVariableSelect, isDebugMode: isDebugMode });}))));
  };
  return ComponentShowcase;
}(React.Component);
var ComponentShowcaseRow = /** @class */function (_super) {
  tslib_1.__extends(ComponentShowcaseRow, _super);
  function ComponentShowcaseRow() {
    var _this = _super !== null && _super.apply(this, arguments) || this;
    _this.getCss = function () {
      return React.createElement("span", { className: ThemeShowcase_module_less_1["default"].relativeCss }, _this.props.row.contents);
    };
    return _this;
  }
  ComponentShowcaseRow.prototype.render = function () {
    var _this = this;
    var _a = this.props,el = _a.element,row = _a.row,isDebugMode = _a.isDebugMode;
    var rowSpan = row.variables.length + 1;
    return React.createElement(React.Fragment, null,
    React.createElement("tr", { className: ThemeShowcase_module_less_1["default"].invisibleRow },
    React.createElement("td", { rowSpan: rowSpan },
    React.createElement(Tooltip_1["default"], { render: this.getCss, pos: 'bottom left', allowedPositions: CSS_TOOLTIP_ALLOWED_POSITIONS, trigger: 'click', useWrapper: false },
    React.createElement("span", { className: ThemeShowcase_module_less_1["default"].elementName },
    ".",
    el))),
    React.createElement("td", { className: ThemeShowcase_module_less_1["default"].invisibleCell }),
    React.createElement("td", { className: ThemeShowcase_module_less_1["default"].invisibleCell }),
    React.createElement("td", { className: ThemeShowcase_module_less_1["default"].invisibleCell })),
    row.variables.map(function (varName) {
      var dependencies = row.dependencies[varName] || EMPTY_ARRAY;
      var variableDefault = DefaultTheme_1["default"][varName];
      var variableFlat = FlatTheme_1["default"][varName];
      var hasNoVariables = isDebugMode && !variableDefault && !variableFlat;
      return React.createElement("tr", { key: el + "_" + varName, className: hasNoVariables ? ThemeShowcase_module_less_1["default"].suspiciousRow : undefined },
      React.createElement("td", null,
      React.createElement(VariableName, { variableName: varName, dependencies: dependencies, onVariableSelect: _this.props.onVariableSelect })),
      React.createElement("td", null,
      React.createElement(VariableValue, { value: variableDefault })),
      React.createElement("td", null,
      React.createElement(VariableValue, { value: variableFlat })));
    }));
  };
  return ComponentShowcaseRow;
}(React.Component);
var VariableName = /** @class */function (_super) {
  tslib_1.__extends(VariableName, _super);
  function VariableName() {
    var _this = _super !== null && _super.apply(this, arguments) || this;
    _this.handleVariableSelect = function () {
      var _a = _this.props,variableName = _a.variableName,onVariableSelect = _a.onVariableSelect;
      if (onVariableSelect) {
        onVariableSelect(event, { value: variableName, label: variableName });
      }
    };
    return _this;
  }
  VariableName.prototype.render = function () {
    return React.createElement("span", null,
    React.createElement("span", { className: ThemeShowcase_module_less_1["default"].variableName, onClick: this.handleVariableSelect }, this.props.variableName),
    this.props.dependencies.length > 0 && this.renderDependencies());
  };
  VariableName.prototype.renderDependencies = function () {
    var _a = this.props,dependencies = _a.dependencies,onVariableSelect = _a.onVariableSelect;
    return React.createElement(React.Fragment, null,
    React.createElement("br", null),
    React.createElement("br", null),
    "\u0437\u0430\u0432\u0438\u0441\u0438\u0442 \u043E\u0442:",
    dependencies.map(function (dependency) {return React.createElement(DependencyName, { key: "dependency_" + dependency, dependencyName: dependency, onDependencySelect: onVariableSelect });}));
  };
  return VariableName;
}(React.Component);
var DependencyName = /** @class */function (_super) {
  tslib_1.__extends(DependencyName, _super);
  function DependencyName() {
    var _this = _super !== null && _super.apply(this, arguments) || this;
    _this.getValues = function () {
      var dependencyName = _this.props.dependencyName;
      var dependencyDefault = DefaultTheme_1["default"][dependencyName];
      var dependencyFlat = FlatTheme_1["default"][dependencyName];
      return React.createElement(React.Fragment, null,
      React.createElement("span", null,
      "Default value: ",
      React.createElement(VariableValue, { value: dependencyDefault })),
      React.createElement("br", null),
      React.createElement("span", null,
      "Flat value: ",
      React.createElement(VariableValue, { value: dependencyFlat })));
    };
    _this.handleDependencySelect = function () {
      var _a = _this.props,dependencyName = _a.dependencyName,onDependencySelect = _a.onDependencySelect;
      if (onDependencySelect) {
        onDependencySelect(event, { value: dependencyName, label: dependencyName });
      }
    };
    return _this;
  }
  DependencyName.prototype.render = function () {
    return React.createElement(React.Fragment, null,
    React.createElement("br", null),
    "\u2013",
    ' ',
    React.createElement(Tooltip_1["default"], { trigger: 'hover', render: this.getValues, pos: 'right middle' },
    React.createElement("span", { className: ThemeShowcase_module_less_1["default"].variableName, onClick: this.handleDependencySelect }, this.props.dependencyName)));
  };
  return DependencyName;
}(React.Component);
var VariableValue = function VariableValue(props) {
  var value = props.value;
  var valueIsColor = isColor(value);
  var valueIsGradient = isGradient(value);
  var hasExample = valueIsColor || valueIsGradient;
  var borderColor = 'transparent';
  if (hasExample) {
    borderColor = valueIsColor ? ColorFunctions_1["default"].contrast(value) : '#000';
  }
  return React.createElement("span", { className: !value ? ThemeShowcase_module_less_1["default"].undefined : undefined },
  hasExample && React.createElement("span", { className: ThemeShowcase_module_less_1["default"].colorExample, style: { background: value, borderColor: borderColor } }),
  value || 'undefined');
};
var ShowUnusedVariables = function ShowUnusedVariables(props) {
  if (props.diff.length === 0) {
    return null;
  }
  return React.createElement("div", { className: ThemeShowcase_module_less_1["default"].unusedVariablesWarning },
  "\u041D\u0435\u0438\u0441\u043F\u043E\u043B\u044C\u0437\u043E\u0432\u0430\u043D\u043D\u044B\u0435 \u043F\u0435\u0440\u0435\u043C\u0435\u043D\u043D\u044B\u0435 (",
  props.diff.length,
  "):",
  React.createElement("ul", null, props.diff.sort().map(function (v) {return React.createElement("li", { key: v }, v);})));
};
function isColor(input) {
  return !!input && (input.startsWith('#') || input.startsWith('rgb') || input.startsWith('hsl'));
}
function isGradient(input) {
  return !!input && input.startsWith('linear-gradient');
}
